File: src/lib/huffman/tree.ts
import type { HuffmanNode } from './types';

export function createNode(
  frequency: number,
  char: string | null = null,
  left: HuffmanNode | null = null,
  right: HuffmanNode | null = null
): HuffmanNode {
  return { frequency, char, left, right };
}

export function isLeaf(node: HuffmanNode): boolean {
  return node.left === null && node.right === null;
}

export function buildHuffmanTree(frequencyMap: Map<string, number>): HuffmanNode | null {
  if (frequencyMap.size === 0) {
    return null;
  }

  if (frequencyMap.size === 1) {
    const [char, freq] = [...frequencyMap.entries()][0];
    const root = createNode(freq);
    root.left = createNode(freq, char);
    return root;
  }

  const nodes: HuffmanNode[] = [];
  frequencyMap.forEach((freq, char) => {
    nodes.push(createNode(freq, char));
  });

  while (nodes.length > 1) {
    nodes.sort((a, b) => a.frequency - b.frequency);

    const left = nodes.shift()!;
    const right = nodes.shift()!;

    const merged = createNode(
      left.frequency + right.frequency,
      null,
      left,
      right
    );

    nodes.push(merged);
  }

  return nodes[0] || null;
}

export function generateCodes(
  root: HuffmanNode | null,
  prefix: string = '',
  codes: Map<string, string> = new Map()
): Map<string, string> {
  if (root === null) {
    return codes;
  }

  if (isLeaf(root) && root.char !== null) {
    codes.set(root.char, prefix || '0');
    return codes;
  }

  if (root.left) {
    generateCodes(root.left, prefix + '0', codes);
  }
  if (root.right) {
    generateCodes(root.right, prefix + '1', codes);
  }

  return codes;
}

export function serializeTree(node: HuffmanNode | null): string {
  if (!node) return '';

  const serialize = (n: HuffmanNode): object => {
    const result: Record<string, unknown> = {
      f: n.frequency,
    };
    if (n.char !== null) {
      result.c = n.char;
    }
    if (n.left) {
      result.l = serialize(n.left);
    }
    if (n.right) {
      result.r = serialize(n.right);
    }
    return result;
  };

  return JSON.stringify(serialize(node));
}

export function deserializeTree(data: string): HuffmanNode | null {
  if (!data) return null;

  const deserialize = (obj: Record<string, unknown>): HuffmanNode => {
    const node = createNode(
      obj.f as number,
      (obj.c as string) ?? null
    );
    if (obj.l) {
      node.left = deserialize(obj.l as Record<string, unknown>);
    }
    if (obj.r) {
      node.right = deserialize(obj.r as Record<string, unknown>);
    }
    return node;
  };

  try {
    const parsed = JSON.parse(data);
    return deserialize(parsed);
  } catch {
    return null;
  }
}

File: backend/huffman/tree.py
import heapq
from typing import Optional
from dataclasses import dataclass, field


@dataclass(order=True)
class HuffmanNode:
    frequency: int
    char: Optional[str] = field(compare=False, default=None)
    left: Optional['HuffmanNode'] = field(compare=False, default=None)
    right: Optional['HuffmanNode'] = field(compare=False, default=None)

    def is_leaf(self) -> bool:
        return self.left is None and self.right is None

    def to_dict(self) -> dict:
        result = {
            'frequency': self.frequency,
            'char': self.char,
            'isLeaf': self.is_leaf()
        }
        if self.left:
            result['left'] = self.left.to_dict()
        if self.right:
            result['right'] = self.right.to_dict()
        return result

    @classmethod
    def from_dict(cls, data: dict) -> 'HuffmanNode':
        node = cls(
            frequency=data['frequency'],
            char=data.get('char')
        )
        if 'left' in data:
            node.left = cls.from_dict(data['left'])
        if 'right' in data:
            node.right = cls.from_dict(data['right'])
        return node


def build_huffman_tree(frequency_map: dict[str, int]) -> Optional[HuffmanNode]:
    if not frequency_map:
        return None

    if len(frequency_map) == 1:
        char, freq = list(frequency_map.items())[0]
        root = HuffmanNode(frequency=freq)
        root.left = HuffmanNode(frequency=freq, char=char)
        return root

    heap = []
    for char, freq in frequency_map.items():
        heapq.heappush(heap, HuffmanNode(frequency=freq, char=char))

    while len(heap) > 1:
        left = heapq.heappop(heap)
        right = heapq.heappop(heap)

        merged = HuffmanNode(
            frequency=left.frequency + right.frequency,
            left=left,
            right=right
        )
        heapq.heappush(heap, merged)

    return heap[0] if heap else None


def generate_codes(root: Optional[HuffmanNode], prefix: str = '', codes: dict = None) -> dict[str, str]:
    if codes is None:
        codes = {}

    if root is None:
        return codes

    if root.is_leaf() and root.char is not None:
        codes[root.char] = prefix if prefix else '0'
        return codes

    if root.left:
        generate_codes(root.left, prefix + '0', codes)
    if root.right:
        generate_codes(root.right, prefix + '1', codes)

    return codes
